client/packages/core/__tests__/src/transactionValidation.test.ts

```diff
@@ -0,0 +1,369 @@
+        has: 'many',
+        label: 'comments',
+      },
+      reverse: {
+        on: 'comments',
+        has: 'one',
+        label: 'post',
+      },
+    },
+    friendships: {
+      forward: {
+        on: 'users',
+        has: 'many',
+        label: 'friends',
+      },
+      reverse: {
+        on: 'users',
+        has: 'many',
+        label: '_friends',
+      },
+    },
+  },
+});
+
+const beValid = (
+  chunk: any,
+  schema: InstantSchemaDef<any, any, any> | null = testSchema,
+) => {
+  expect(() => validateTransactions(chunk, schema ?? undefined)).not.toThrow();
+  if (schema) {
+    expect(() => validateTransactions(chunk, undefined)).not.toThrow();
+  }
+};
+
+const beWrong = (
+  chunk: any,
+  schema: InstantSchemaDef<any, any, any> | null = testSchema,
+) => {
+  expect(() => validateTransactions(chunk, schema ?? undefined)).toThrow();
+};
+
+const tx = originalTx as unknown as TxChunk<typeof testSchema>;
+
+test('validates basic transaction chunk', () => {
+  const userId = id();
+  const validChunk = tx.users[userId].create({
+    name: 'John',
+    email: 'john@example.com',
+  });
+
+  beValid(validChunk);
+});
+
+test('validates transaction chunk arrays', () => {
+  const userId = id();
+  const postId = id();
+  const chunks = [
+    tx.users[userId].create({ name: 'John', email: 'john@example.com' }),
+    tx.posts[postId].create({ title: 'Hello', body: 'World' }),
+  ];
+
+  beValid(chunks);
+});
+
+test('validates create operations', () => {
+  const userId = id();
+
+  // Valid create
+  beValid(tx.users[userId].create({ name: 'John', email: 'john@example.com' }));
+
+  // Valid create with optional field
+  beValid(
+    tx.users[userId].create({
+      name: 'John',
+      email: 'john@example.com',
+      bio: 'Developer',
+    }),
+  );
+
+  // Valid create with any type
+  beValid(
+    tx.users[userId].create({
+      name: 'John',
+      email: 'john@example.com',
+      junk: { anything: 'goes' },
+    }),
+  );
+
+  // Invalid create - wrong type
+  // @ts-expect-error
+  beWrong(tx.users[userId].create({ name: 123, email: 'john@example.com' }));
+
+  // Invalid create - unknown attribute
+  beWrong(
+    tx.users[userId].create({
+      name: 'John',
+      email: 'john@example.com',
+      // @ts-expect-error
+      unknownField: 'value',
+    }),
+  );
+
+  // Invalid create - non-object args
+  beWrong({
+    __ops: [['create', 'users', userId, 'not an object']],
+    __etype: 'users',
+  });
+});
+
+test('validates update operations', () => {
+  const userId = id();
+
+  // Valid update
+  beValid(tx.users[userId].update({ name: 'Jane' }));
+
+  // Valid update with multiple fields
+  beValid(tx.users[userId].update({ name: 'Jane', bio: 'Updated bio' }));
+
+  // Invalid update - wrong type
+  // @ts-expect-error
+  beWrong(tx.users[userId].update({ name: 123 }));
+
+  // Invalid update - unknown attribute
+  // @ts-expect-error
+  beWrong(tx.users[userId].update({ unknownField: 'value' }));
```

Comment:
A part of me hesitates in erroring out when we receive extra fields. It feels unrichhickey-like -- but thinking it through I do think it makes sense. Wanted to flag in case others disagree with me. @nezaj @tonsky @dwwoelfel